data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconBlockHandlerTest.java

```diff
@@ -0,0 +1,299 @@
+
+    final String jsonResponse =
+        jsonProvider.objectToJSON(new BeaconBlockResponse(signedBeaconBlock));
+    verify(context).result(jsonResponse);
+  }
+
+  @Test
+  public void shouldReturnBlockWhenSlotQuery() throws Exception {
+    final String slotNum = "1";
+    final Map<String, List<String>> params = Map.of("slot", List.of(slotNum));
+
+    when(context.queryParamMap()).thenReturn(params);
+    when(context.queryParam(any())).thenReturn(slotNum);
+    when(storageClient.getStore()).thenReturn(store);
+    when(storageClient.getBlockRootBySlot(any())).thenReturn(Optional.of(blockRoot));
+    when(store.getSignedBlock(any())).thenReturn(signedBeaconBlock);
+    when(historicalChainData.getFinalizedBlockAtSlot(any()))
+        .thenReturn(SafeFuture.completedFuture(empty()));
+
+    handler.handle(context);
+
+    final String jsonResponse =
+        jsonProvider.objectToJSON(new BeaconBlockResponse(signedBeaconBlock));
+    verify(context).result(jsonResponse);
+  }
+
+  @Test
+  public void shouldReturnBlockWhenEpochQueryAndNoBlockRoot() throws Exception {
+    final String epochNum = "1";
+    final Map<String, List<String>> params = Map.of("epoch", List.of(epochNum));
+
+    when(context.queryParamMap()).thenReturn(params);
+    when(context.queryParam(any())).thenReturn(epochNum);
+    when(storageClient.getBlockRootBySlot(any())).thenReturn(empty());
+    when(historicalChainData.getFinalizedBlockAtSlot(any()))
+        .thenReturn(SafeFuture.completedFuture(Optional.of(signedBeaconBlock)));
+
+    handler.handle(context);
+
+    final String jsonResponse =
+        jsonProvider.objectToJSON(new BeaconBlockResponse(signedBeaconBlock));
+    verify(context).result(jsonResponse);
+  }
+
+  @Test
+  public void shouldReturnBlockWhenSlotQueryAndNoBlockRoot() throws Exception {
+    final String slotNum = "1";
+    final Map<String, List<String>> params = Map.of("slot", List.of(slotNum));
+
+    when(context.queryParamMap()).thenReturn(params);
+    when(context.queryParam(any())).thenReturn(slotNum);
+    when(storageClient.getBlockRootBySlot(any())).thenReturn(empty());
+    when(historicalChainData.getFinalizedBlockAtSlot(any()))
+        .thenReturn(SafeFuture.completedFuture(Optional.of(signedBeaconBlock)));
+
+    handler.handle(context);
+
+    final String jsonResponse =
+        jsonProvider.objectToJSON(new BeaconBlockResponse(signedBeaconBlock));
+    verify(context).result(jsonResponse);
+  }
+
+  @Test
+  public void shouldFailWhenNoParams() throws Exception {
+    handler.handle(realContext);
+
+    final String actualResponse = realContext.resultString();
+    final String expectedResponse = jsonProvider.objectToJSON(BAD_REQUEST);
+
+    assertThat(actualResponse).isEqualTo(expectedResponse);
+  }
+
+  @Test
+  public void shouldFailWithEmptyRootParamValue() throws Exception {
+    final Map<String, List<String>> params = Map.of("root", emptyList());
+
+    doReturn(params).when(realContext).queryParamMap();
```

Comment:
probably should try to be consistent with the pattern, either when.thenReturn or doReturn.when...

Reply:
`realContext` is a spy(), which (IIRC) requires the doReturn pattern due to evaluation order.